orchagent/fgnhgorch.cpp

```diff
@@ -0,0 +1,1500 @@
+    {
+        bank_index_range bir;
+        bir.start_index = prev_idx;
+        bir.end_index = bir.start_index + (buckets_per_nexthop * memb_per_bank[i]) + split_extra_buckets_among_bank - 1;
+        if (extra_buckets > 0)
+        {
+            bir.end_index = bir.end_index + 1;
+            extra_buckets--;
+        }
+        if (i == fgNhgEntry->hash_bucket_indices.size())
+        {
+            fgNhgEntry->hash_bucket_indices.push_back(bir);
+        }
+        else
+        {
+            fgNhgEntry->hash_bucket_indices[i] = bir;
+        }
+        prev_idx = bir.end_index + 1;
+        SWSS_LOG_INFO("Calculate_bank_hash_bucket_start_indices: bank %d, si %d, ei %d",
+                       i, fgNhgEntry->hash_bucket_indices[i].start_index, fgNhgEntry->hash_bucket_indices[i].end_index);
+    }
+}
+
+
+void FgNhgOrch::set_state_db_route_entry(const IpPrefix &ipPrefix, uint32_t index, NextHopKey nextHop)
+{
+    SWSS_LOG_ENTER();
+
+    string key = ipPrefix.to_string();
+    // Write to StateDb
+    std::vector<FieldValueTuple> fvs;
+
+    // check if profile already exists - if yes - skip creation
+    m_stateWarmRestartRouteTable.get(key, fvs);
+
+    //bucket rewrite
+    if (fvs.size() > index)
+    {
+        FieldValueTuple fv(std::to_string(index), nextHop.to_string());
+        fvs[index] = fv;
+        SWSS_LOG_INFO("Set state db entry for ip prefix %s next hop %s with index %d",
+                        ipPrefix.to_string().c_str(), nextHop.to_string().c_str(), index);
+        m_stateWarmRestartRouteTable.set(key, fvs);
+    }
+    else
+    {
+        fvs.push_back(FieldValueTuple(std::to_string(index), nextHop.to_string()));
+        SWSS_LOG_INFO("Add new next hop entry %s with index %d for ip prefix %s",
+                nextHop.to_string().c_str(), index, ipPrefix.to_string().c_str());
+        m_stateWarmRestartRouteTable.set(key, fvs);
+    }
+
+}
+
+
+bool FgNhgOrch::write_hash_bucket_change(FGNextHopGroupEntry *syncd_fg_route_entry, uint32_t index, sai_object_id_t nh_oid,
+        const IpPrefix &ipPrefix, NextHopKey nextHop)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t nhgm_attr;
+    nhgm_attr.id = SAI_NEXT_HOP_GROUP_MEMBER_ATTR_NEXT_HOP_ID;
+    nhgm_attr.value.oid = nh_oid;
+    sai_status_t status = sai_next_hop_group_api->set_next_hop_group_member_attribute(
+                                                              syncd_fg_route_entry->nhopgroup_members[index],
+                                                              &nhgm_attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to set next hop oid %" PRIx64 " member %" PRIx64 ": %d",
+            syncd_fg_route_entry->nhopgroup_members[index], nh_oid, status);
+        return false;
+    }
+
+    set_state_db_route_entry(ipPrefix, index, nextHop);
+    return true;
+}
+
+
+bool FgNhgOrch::create_fine_grained_next_hop_group(FGNextHopGroupEntry &syncd_fg_route_entry, FgNhgEntry *fgNhgEntry,
```

Comment:
>create_fine_grained_next_hop_group [](start = 16, length = 34)

Is it possible to move this function into class RouteOrch? Then no need to expose XxxNextHopGroupCount() functions.

Reply:
Moved it to RouteOrch